# CO2 Emissions Analysis

In this project, I try to analyze the CO2 and Greenhouse Gases Emissions data provided by *Our World in Data*. The descriptions for the columns used in this analysis are as follows:

* **country**              : Geographic location.	
* **year**                 : Year of observation.	
* **iso_code**             : ISO 3166-1 alpha-3, three-letter country codes.
* **co2**                  : Annual production-based emissions of carbon dioxide (CO2), measured in million tonnes.
* **coal_co2**             : Annual production-based emissions of carbon dioxide (CO2) from coal, measured in million tonnes.
* **gas_co2**              : Annual production-based emissions of carbon dioxide (CO2) from gas, measured in million tonnes.
* **oil_co2**              : Annual production-based emissions of carbon dioxide (CO2) from oil, measured in million tonnes.
* **share_global_co2**     : Annual production-based emissions of carbon dioxide (CO2), measured as a percentage of global       production-based emissions of CO2 in the same year.
* **share_global_cumulative_co2**: Cumulative production-based emissions of carbon dioxide (CO2) since the first year of data availability, measured as a percentage of global cumulative production-based emissions of CO2 since the first year of data availability.


The full list of columns and their descriptions can be found [here](https://github.com/owid/co2-data/blob/master/owid-co2-codebook.csv).

## Imports

In [26]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly
import cufflinks as cf
import plotly.graph_objects as go
cf.go_offline()
%matplotlib inline

## Getting the Data

In [2]:
df = pd.read_csv('owid-co2-data.csv')

In [3]:
df.head()

country  year iso_code  population           gdp  cement_co2  \
0  Afghanistan  1949      AFG   7624058.0           NaN         NaN   
1  Afghanistan  1950      AFG   7752117.0  9.421400e+09         NaN   
2  Afghanistan  1951      AFG   7840151.0  9.692280e+09         NaN   
3  Afghanistan  1952      AFG   7935996.0  1.001733e+10         NaN   
4  Afghanistan  1953      AFG   8039684.0  1.063052e+10         NaN   

   cement_co2_per_capita    co2  co2_growth_abs  co2_growth_prct  ...  \
0                    NaN  0.015             NaN              NaN  ...   
1                    NaN  0.084           0.070            475.0  ...   
2                    NaN  0.092           0.007              8.7  ...   
3                    NaN  0.092           0.000              0.0  ...   
4                    NaN  0.106           0.015             16.0  ...   

   share_global_cumulative_oil_co2  share_global_cumulative_other_co2  \
0                              NaN                                NaN   
1                              0.0                                NaN   
2                              0.0                                NaN   
3                              0.0                                NaN   
4                              0.0                                NaN   

   share_global_flaring_co2  share_global_gas_co2  share_global_oil_co2  \
0                       NaN                   NaN                   NaN   
1                       NaN                   NaN                   0.0   
2                       NaN                   NaN                   0.0   
3                       NaN                   NaN                   0.0   
4                       NaN                   NaN                   0.0   

   share_global_other_co2  total_ghg  total_ghg_excluding_lucf  trade_co2  \
0                     NaN        NaN                       NaN        NaN   
1                     NaN        NaN                       NaN        NaN   
2                     NaN        NaN                       NaN        NaN   
3                     NaN        NaN                       NaN        NaN   
4                     NaN        NaN                       NaN        NaN   

   trade_co2_share  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  

[5 rows x 60 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26008 entries, 0 to 26007
Data columns (total 60 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   country                              26008 non-null  object 
 1   year                                 26008 non-null  int64  
 2   iso_code                             21913 non-null  object 
 3   population                           23878 non-null  float64
 4   gdp                                  13479 non-null  float64
 5   cement_co2                           12668 non-null  float64
 6   cement_co2_per_capita                12638 non-null  float64
 7   co2                                  24670 non-null  float64
 8   co2_growth_abs                       24294 non-null  float64
 9   co2_growth_prct                      25696 non-null  float64
 10  co2_per_capita                       24032 non-null  float64
 11  co2_per_gdp                 

In [5]:
df.describe()

year    population           gdp    cement_co2  \
count  26008.000000  2.387800e+04  1.347900e+04  12668.000000   
mean    1952.169525  9.336274e+07  2.887445e+11     15.853638   
std       54.562304  4.066021e+08  2.184803e+12     84.179826   
min     1750.000000  1.490000e+03  5.543200e+07      0.000000   
25%     1923.000000  1.401949e+06  9.759778e+09      0.134000   
50%     1966.000000  5.288000e+06  3.041914e+10      0.603000   
75%     1994.000000  2.234403e+07  1.273866e+11      3.255000   
max     2020.000000  7.794789e+09  1.136302e+14   1626.371000   

       cement_co2_per_capita           co2  co2_growth_abs  co2_growth_prct  \
count           12638.000000  24670.000000    24294.000000     25696.000000   
mean                0.111586    326.658348        6.383185        20.760047   
std                 0.147532   1677.027132       61.526038       692.088419   
min                 0.000000      0.000000    -1895.244000       -99.640000   
25%                 0.019000      0.557000       -0.007000        -0.332500   
50%                 0.068000      5.333000        0.066000         3.370000   
75%                 0.155000     48.153250        1.242000        10.380000   
max                 2.738000  36702.504000     1736.258000    102318.508000   

       co2_per_capita   co2_per_gdp  ...  share_global_cumulative_oil_co2  \
count    24032.000000  15851.000000  ...                     21100.000000   
mean         4.115845      0.428189  ...                         3.684855   
std         14.700555      0.480043  ...                        13.601320   
min          0.000000      0.000000  ...                         0.000000   
25%          0.253000      0.145000  ...                         0.010000   
50%          1.226000      0.285000  ...                         0.080000   
75%          4.612250      0.544000  ...                         0.630000   
max        748.639000      7.718000  ...                       100.000000   

       share_global_cumulative_other_co2  share_global_flaring_co2  \
count                        2208.000000               4641.000000   
mean                           17.432921                  6.870909   
std                            30.325683                 15.449948   
min                             0.000000                  0.000000   
25%                             0.230000                  0.110000   
50%                             1.105000                  0.790000   
75%                            18.762500                  5.780000   
max                           100.000000                100.000000   

       share_global_gas_co2  share_global_oil_co2  share_global_other_co2  \
count           9245.000000          21100.000000             2208.000000   
mean               6.664776              3.669233               18.240177   
std               20.434607             13.393958               31.649508   
min                0.000000              0.000000                0.000000   
25%                0.030000              0.010000                0.340000   
50%                0.220000              0.090000                1.550000   
75%                1.500000              0.650000               19.362500   
max              100.000000            100.000000              100.000000   

          total_ghg  total_ghg_excluding_lucf    trade_co2  trade_co2_share  
count   6149.000000               6149.000000  4096.000000      4096.000000  
mean     771.485169                748.578503    -8.124469        21.584172  
std     3553.425128               3488.004831   262.090604        45.459095  
min     -186.550000                  0.010000 -2233.000000       -96.760000  
25%        8.440000                  7.030000    -1.663750        -2.857500  
50%       38.050000                 30.600000     1.874500        10.790000  
75%      151.150000                131.320000     9.700500        35.017500  
max    49758.230000              48116.559000  2047.575000       366.15000

## Exploratory Data Analysis

In [6]:
df.isnull().sum()

country                                    0
year                                       0
iso_code                                4095
population                              2130
gdp                                    12529
cement_co2                             13340
cement_co2_per_capita                  13370
co2                                     1338
co2_growth_abs                          1714
co2_growth_prct                          312
co2_per_capita                          1976
co2_per_gdp                            10157
co2_per_unit_energy                    16050
coal_co2                                8099
coal_co2_per_capita                     8472
consumption_co2                        21912
consumption_co2_per_capita             21912
consumption_co2_per_gdp                22131
cumulative_cement_co2                  13340
cumulative_co2                          1338
cumulative_coal_co2                     8099
cumulative_flaring_co2                 21367
cumulative

There seems to be a lot of missing values. We can filter out the first quartile for 'year' since the data starts from 1750 and most likely wouldn't contain a lot of information. For the purpose of this analysis, we will restrict the analysis to the time period 1920 to 2020. 

In [7]:
df[df['year']>1920].isnull().sum()

country                                    0
year                                       0
iso_code                                2232
population                              1245
gdp                                     7930
cement_co2                              7243
cement_co2_per_capita                   7273
co2                                      700
co2_growth_abs                           886
co2_growth_prct                          220
co2_per_capita                          1010
co2_per_gdp                             6508
co2_per_unit_energy                     9737
coal_co2                                7378
coal_co2_per_capita                     7423
consumption_co2                        15599
consumption_co2_per_capita             15599
consumption_co2_per_gdp                15818
cumulative_cement_co2                   7243
cumulative_co2                           700
cumulative_coal_co2                     7378
cumulative_flaring_co2                 15054
cumulative

In [8]:
df = df[df['year']>1920]

In [9]:
df.shape

(19695, 60)

We can drop the columns with more than 80% missing values

In [10]:
thresh = len(df) * 0.2
df.dropna(axis=1,thresh=thresh,inplace=True)

In [11]:
df.shape

(19695, 54)

In [12]:
df.columns

Index(['country', 'year', 'iso_code', 'population', 'gdp', 'cement_co2',
       'cement_co2_per_capita', 'co2', 'co2_growth_abs', 'co2_growth_prct',
       'co2_per_capita', 'co2_per_gdp', 'co2_per_unit_energy', 'coal_co2',
       'coal_co2_per_capita', 'consumption_co2', 'consumption_co2_per_capita',
       'cumulative_cement_co2', 'cumulative_co2', 'cumulative_coal_co2',
       'cumulative_flaring_co2', 'cumulative_gas_co2', 'cumulative_oil_co2',
       'energy_per_capita', 'energy_per_gdp', 'flaring_co2',
       'flaring_co2_per_capita', 'gas_co2', 'gas_co2_per_capita',
       'ghg_excluding_lucf_per_capita', 'ghg_per_capita', 'methane',
       'methane_per_capita', 'nitrous_oxide', 'nitrous_oxide_per_capita',
       'oil_co2', 'oil_co2_per_capita', 'primary_energy_consumption',
       'share_global_cement_co2', 'share_global_co2', 'share_global_coal_co2',
       'share_global_cumulative_cement_co2', 'share_global_cumulative_co2',
       'share_global_cumulative_coal_co2',
       's

### Global CO2 Emissions Trend

In [59]:
worlddf = df[df['country']=='World']
fig = go.Figure()

fig.add_traces(go.Scatter(x=worlddf['year'], y=worlddf['co2'], mode='lines', showlegend=False))

fig.update_layout(title='Global CO2 Emissions')
fig.update_xaxes(title='Year')
fig.update_yaxes(title='CO2 Emissions (in million tonnes)')

Plotting markers on the troughs

In [60]:
newdf = pd.DataFrame(worlddf[['year','co2']])
def get_markers(co2,year):
    if year in [1992,2009,2020]:
        return co2
    else:
        return np.NaN
newdf['co2'] = newdf.apply(lambda x: get_markers(x.co2,x.year), axis=1)

In [61]:
fig = go.Figure()

fig.add_traces(go.Scatter(x=worlddf['year'], y=worlddf['co2'], mode='lines', showlegend=False))

fig.add_traces(go.Scatter(x=newdf['year'], y=newdf['co2'], mode='markers', marker=dict(color='red',symbol='square'), showlegend=False))

fig.update_layout(title='Global CO2 Emissions')
fig.update_xaxes(title='Year')
fig.update_yaxes(title='CO2 Emissions (in million tonnes)')

**Interestingly, the low points coincide with economic crises and major events - COVID 19 pandemic in 2020, global financial crisis and oil crisis in 2008, Soviet Union collapse, Gulf war and Indian economic crisis in 1991** 

In [62]:
df[df['country']=='World'].iplot(x='year',y=['co2','gas_co2','oil_co2','coal_co2'],title='Global CO2 Emissions due to Gas, Oil and Coal',
                                 xTitle='Year',yTitle='CO2 Emissions (in million tonnes)')

### Analyzing top contributors to global emissions

In [15]:
df[df['year']==2020][['country','co2']].sort_values(by='co2',ascending=False).head(10)

country        co2
25747                          World  34807.258
1250                            Asia  20317.059
24871  Upper-middle-income countries  15895.922
10662          High-income countries  11828.815
4730                           China  10667.888
1441      Asia (excl. China & India)   7207.378
14146  Lower-middle-income countries   5825.887
17018                  North America   5775.159
7279                          Europe   4946.035
24680                  United States   4712.771

The dataset contains the data for categories like continents, income categories etc. These records will not have iso codes as they are not countries. We create a list of entities that are not countries so that we can exclude them if needed.

In [16]:
entitylist = df[df['iso_code'].isnull()==True]['country'].unique().tolist()

In [17]:
entitylist

['Africa',
 'Asia',
 'Asia (excl. China & India)',
 'Europe',
 'Europe (excl. EU-27)',
 'Europe (excl. EU-28)',
 'European Union (27)',
 'European Union (28)',
 'French Equatorial Africa',
 'French West Africa',
 'High-income countries',
 'International transport',
 'Kosovo',
 'Kuwaiti Oil Fires',
 'Leeward Islands',
 'Low-income countries',
 'Lower-middle-income countries',
 'North America',
 'North America (excl. USA)',
 'Oceania',
 'Panama Canal Zone',
 'Ryukyu Islands',
 'South America',
 'St. Kitts-Nevis-Anguilla',
 'Upper-middle-income countries',
 'World']

We identify the top 10 countries that contributed to CO2 emissions as of 2020 and plot their trend over the last 100 years.

In [18]:
top10co2 = df[(~df['country'].isin(entitylist)) & (df['year']==2020)][['country','co2']].sort_values(by='co2',ascending=False).head(10)
top10 = top10co2['country'].to_list()

In [19]:
plotly.express.line(data_frame=df[df['country'].isin(top10)], x='year', y='co2', line_group='country',
                    color='country', title='Annual CO2 Emissions by Country', category_orders={'country':top10},
                    labels={'year':'Year','co2':'Annual CO2 Emissions (in million tonnes)','country':'Country'})

**Russis's sudden decline in emissions since 1991 may be attributed to the collapse of the Soviet Union.**

**The increase of CO2 emissions in China through out the 2000's may be attributed to its fast economic growth during the time and focus on manufacturing**

### Global Share of the top 10 countries 

In [20]:
pie_df = df[(df['country'].isin(top10)) & (df['year']==2020)]

Filtering out the columns that do not contain global share details (except 'country') in the new dataframe

In [21]:
filter_col = [col for col in df.columns if (col.startswith('share')!=True)]
filter_col.remove('country')
pie_df.drop(filter_col,axis=1,inplace=True)

Adding a row 'Rest of the World' to represent the rest of the countries

In [22]:
newdf = pd.DataFrame(columns=pie_df.columns)
newdf['country']=['Rest of the World']
for x in pie_df.columns:
    if x == 'country':
        continue 
    newdf[x] = [100 - pie_df[x].sum()]
pie_df = pd.concat([pie_df,newdf])
top10.append('Rest of the World')

The dataframe to plot the pie chart

In [23]:
pie_df.head(11)

country  share_global_cement_co2  share_global_co2  \
4730               China                    52.77             30.65   
9597             Germany                     0.82              1.85   
11255              India                     7.56              7.02   
11387          Indonesia                     2.08              1.69   
11573               Iran                     1.47              2.14   
12314              Japan                     1.56              2.96   
19595             Russia                     1.25              4.53   
20268       Saudi Arabia                     1.57              1.80   
21589        South Korea                     1.41              1.72   
24680      United States                     2.51             13.54   
0      Rest of the World                    27.00             32.10   

       share_global_coal_co2  share_global_cumulative_cement_co2  \
4730                   53.10                               34.30   
9597                    1.42                                2.76   
11255                  11.36                                5.73   
11387                   2.15                                1.56   
11573                   0.04                                1.20   
12314                   2.88                                4.59   
19595                   2.55                                3.79   
20268                    NaN                                1.23   
21589                   2.02                                2.17   
24680                   6.36                                6.36   
0                      18.12                               36.31   

       share_global_cumulative_co2  share_global_cumulative_coal_co2  \
4730                         13.89                             22.48   
9597                          5.46                              8.17   
11255                         3.21                              4.47   
11387                         0.85                              0.47   
11573                         1.11                              0.03   
12314                         3.87                              3.01   
19595                         6.80                              5.72   
20268                         0.94                               NaN   
21589                         1.08                              1.02   
24680                        24.56                             22.28   
0                            38.23                             32.35   

       share_global_cumulative_flaring_co2  share_global_cumulative_gas_co2  \
4730                                   NaN                             2.49   
9597                                  0.52                             2.86   
11255                                 0.93                             0.96   
11387                                 2.77                             0.99   
11573                                10.84                             2.86   
12314                                 0.09                             2.79   
19595                                 6.53                            13.93   
20268                                 5.07                             1.94   
21589                                  NaN                             0.80   
24680                                12.67                            31.22   
0                                    60.58                            39.16   

       share_global_cumulative_oil_co2  share_global_flaring_co2  \
4730                              5.52                       NaN   
9597                              3.31                      0.46   
11255                             2.39                      0.39   
11387                             1.19                      2.55   
11573                             1.56                     13.84   
12314                             5.53                      0.09   
19595                             5.52                   

In [24]:
fig = plotly.subplots.make_subplots(rows=1,cols=2,
                                    subplot_titles=['Percentage of Global Emissions','Percentage of Cumulative Global Emisssions'],
                                    specs=[[{'type': 'pie'}, {'type': 'pie'}]])

fig.add_trace(go.Pie(values=pie_df['share_global_co2'], labels=pie_df['country'], 
                     domain=dict(x=[0,0.5]), name='% of Emissions'), row=1, col=1)

fig.add_trace(go.Pie(values=pie_df['share_global_cumulative_co2'], labels=pie_df['country'], 
                     domain=dict(x=[0.5,1.0]), name='% of Cumulative Emissions'), row=1, col=2)



**China and India's share of global emissions in 2020 is more than their share of cumulative emissions - China's being 16 points ahead. United States on the other hand had a lower share in 2020 compared to their share of cumulative emissions**